modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -145,13 +145,153 @@ class SoftMaxSubgraph : public Subgraph
     int axis;
 };
 
+class UpsampleSubgraph : public Subgraph
+{
+public:
+    UpsampleSubgraph()
+    {
+        int input = addNodeToMatch("");
+
+        int indexH = addNodeToMatch("Constant");
+        int inpShape1 = addNodeToMatch("Shape", input);
+        int inpGather1 = addNodeToMatch("Gather", inpShape1, indexH);
+        int inpCastG1 = addNodeToMatch("Cast", inpGather1);
+        int scaleH = addNodeToMatch("Constant");
+        int inpMul1 = addNodeToMatch("Mul", inpCastG1, scaleH);
+        int inpCastM1 = addNodeToMatch("Cast", inpMul1);
+        int inpFloor1 = addNodeToMatch("Floor", inpCastM1);
+
+        int indexW = addNodeToMatch("Constant");
+        int inpShape2 = addNodeToMatch("Shape", input);
+        int inpGather2 = addNodeToMatch("Gather", inpShape2, indexW);
+        int inpCastG2 = addNodeToMatch("Cast", inpGather2);
+        int scaleW = addNodeToMatch("Constant");
+        int inpMul2 = addNodeToMatch("Mul", inpCastG2, scaleW);
+        int inpCastM2 = addNodeToMatch("Cast", inpMul2);
+        int inpFloor2 = addNodeToMatch("Floor", inpCastM2);
+
+        int inpUnsqueeze1 = addNodeToMatch("Unsqueeze", inpFloor1);
+        int inpUnsqueeze2 = addNodeToMatch("Unsqueeze", inpFloor2);
+        int inpConcat1 = addNodeToMatch("Concat", inpUnsqueeze1, inpUnsqueeze2);
+
+        int scaleNC = addNodeToMatch("Constant");
+
+        int inpCast3 = addNodeToMatch("Cast", inpConcat1);
+
+        int inpShape3 = addNodeToMatch("Shape", input);
+        int inpSlice = addNodeToMatch("Slice", inpShape3);
+        int inpCast4 = addNodeToMatch("Cast", inpSlice);
+
+        int inpDiv = addNodeToMatch("Div", inpCast3, inpCast4);
+        int inpConcat2 = addNodeToMatch("Concat", scaleNC, inpDiv);
+
+        addNodeToMatch("Upsample", input, inpConcat2);
+        setFusedNode("Upsample", input, scaleH, scaleW, scaleNC);
+    }
+
+    virtual void finalize(const Ptr<ImportGraphWrapper>&,
+                          const Ptr<ImportNodeWrapper>& fusedNode,
+                          std::vector<Ptr<ImportNodeWrapper> >& inputs) CV_OVERRIDE
+    {
+        opencv_onnx::NodeProto* node = fusedNode.dynamicCast<ONNXNodeWrapper>()->node;
+        CV_Assert(node->input_size() == 4);
+
+        std::vector<opencv_onnx::NodeProto*> input_nodes(3);
+        std::vector<cv::Mat> nodes_attributes(3);
+        input_nodes[0] = inputs[3].dynamicCast<ONNXNodeWrapper>()->node; //scaleNC
+        input_nodes[1] = inputs[1].dynamicCast<ONNXNodeWrapper>()->node; //scaleH
+        input_nodes[2] = inputs[2].dynamicCast<ONNXNodeWrapper>()->node; //scaleW
```

Comment:
I think that we need to check which scale relates to width and which to height (might be swapped). So the axes values are in Gather nodes. I think it's easier to extract values from them at `match` stage and then set in `finalize`. BTW, we can do the same for scales and perform assertions for batch and channels indices.

Reply:
```
virtual bool match(const Ptr<ImportGraphWrapper>& net, int nodeId,
                       std::vector<int>& matchedNodesIds,
                       std::vector<int>& targetNodesIds)
```
We are extrating values from `Costant` nodes.
And while putting ids in `matchedNodeIds/targetNodeIds` constant nodes were skipped. 
https://github.com/opencv/opencv/blob/f3237fdc6e7eb3a0260e0942af27b9fadc9a11ef/modules/dnn/src/graph_simplifier.cpp#L114-L123
That is why I added them to ` inputs` of fused node so that we can extract the values later. I am not sure how to get id of constant node in match function.

Reply:
Let me add check for axes swap.

Reply:
@dkurt , what to do about this ?
```
Total tests: 8605
Tests failed: 0
Tests passed: 8605
Exception has been thrown. Please see stdio log.
```
https://pullrequest.opencv.org/buildbot/builders/precommit_macosx/builds/23190/steps/test_imgproc/logs/tests%20summary

Reply:
@ashishkrshrivastava Please ignore MacOSX builder failures. They are not related to this PR.

Reply:
okay.

Reply:
@ashishkrshrivastava, Maybe we can combine it in this way?
```cpp
if (!Subgraph::match(net, nodeId, matchedNodesIds, targetNodesIds))
    return false;

// Get Constant nodes from "net" using "matchedNodesIds"

return true;
```

in example, from TensorFlow's graphs fusion:

https://github.com/opencv/opencv/blob/4cb9faf6c9c012d603217bc09313d325bb9fb04a/modules/dnn/src/tensorflow/tf_graph_simplifier.cpp#L254-L263

So you can extract scales from there and keep only actual inputs for the fused node

Reply:
ohh, okay. got it.